File: linkedlists/doublyLinkedList.py
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None
        self.prev = None



class doublyLinkedList:
    def __init__(self, value):
        newNode = Node(value)
        self.head = newNode
        self.tail = newNode
        self.length = 1

    def append(self, value):
        newNode = Node(value)
        newNode.prev=self.tail
        self.tail.next = newNode
        self.tail = newNode
        self.length += 1
        return self

    def prepend(self, value):
        newNode = Node(value)
        newNode.next = self.head
        self.head.prev = newNode
        self.head = newNode
        self.length += 1
        return self
    
    def printList(self):
        array = []
        currentNode = self.head
        while currentNode:
            array.append(currentNode.value)
            currentNode = currentNode.next
        return array

    def traverse(self, index):
        currentNode = self.head
        for i in range(index):
            currentNode = currentNode.next
        return currentNode
    
    def insert(self, index, value):
        if index >= self.length:
            return self.append(value)
        elif index<=0:
            return self.prepend(value)
        newNode = Node(value)
        leader = self.traverse(index-1)
        follower = leader.next
        leader.next = newNode
        newNode.prev = leader
        newNode.next = follower
        self.length+=1
        return self

        # currentNode = self.head
        # for i in range(index):
        #     if i == index-1:
        #         newNode.next = currentNode.next
        #         currentNode.next=newNode
        #     else:
        #         currentNode=currentNode.next

    def remove(self, index):
        leader = self.traverse(index-1)
        follower = self.traverse(index+1)
        leader.next = follower
        follower.prev = leader
        self.length -=1
                

LL = doublyLinkedList(10)
LL.append(5)
LL.append(16)
LL.prepend(1)
LL.insert(2, 70)
LL.remove(3)
print(LL.printList())

File: linkedlists/doublyLinkedList.js
class Node {
    constructor(value) {
        this.value = value;
        this.next = null;
        this.prev = null;
    }
}

class linkedList {
    constructor(value) {
        const newNode = new Node(value);
        this.head = newNode;
        this.tail = newNode;
        this.length = 1;
    }

    append(value) {
        const newNode = new Node(value);
        newNode.prev = this.tail;
        this.tail.next = newNode;
        this.tail = newNode;
        this.length++;
        return this;
    }

    prepend(value) {
        const newNode = new Node(value);
        newNode.next = this.head;
        this.head.prev = newNode
        this.head = newNode;
        this.length++;
        return this;
    }

    printList() {
        const array = [];
        let currentNode = this.head;
        while(currentNode != null) {
            array.push(currentNode.value);
            currentNode = currentNode.next;
        }
        return array;
    }

    traverse(index) {
        let counter = 0;
        let currentNode = this.head;
        while (counter !== index) {
            currentNode = currentNode.next;
            counter++
        }
        return currentNode;
    }

    insert(index, value) {
        if (index>=this.length) {
            return this.append(value);
        }
        else if (index <= 0) {
            return this.prepend(value);
        }
        let newNode = new Node(value);
        let leader = this.traverse(index-1);
        let follower = leader.next;
        leader.next = newNode;
        newNode.prev = leader;
        newNode.next = follower;
        this.length++;
        return this;
    }

    remove(index) {
        if (index <= 0) {
            // Remove head
            this.head = this.head.next;
            if (this.head) this.head.prev = null;
            this.length--;
            return this;
        }
        if (index >= this.length - 1) {
            // Remove tail
            this.tail = this.tail.prev;
            if (this.tail) this.tail.next = null;
            this.length--;
            return this;
        }
        const leader = this.traverse(index - 1);
        const unwantedNode = leader.next;
        const follower = unwantedNode.next;
        leader.next = follower;
        follower.prev = leader;
        this.length--;
        return this;
    }
}

LL = new linkedList(10)
LL.append(5)
LL.append(16)
LL.prepend(1)
LL.insert(2,70)
LL.insert(99, 4)
 LL.remove(2)
console.log(LL.printList())
// LL.remove(3)
// print(LL.printList())